### PR TITLE
Add teleport-challenge app scaffolding

### PR DESCRIPTION
**Pull request by jacobboykin**
* Create initial app scaffolding to build on
* Add `/healthcheck` endpoint
* Add binary build and lint CI
* Add documentation for the current state of the app

### CODE REVIEW

File: internal/kubernetes/kubernetes.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+// TODO(time saver): implement a mock client using "k8s.io/client-go/kubernetes/fake" or mockery for testing
+
+type KubeClientController struct {
+	client *kubernetes.Clientset
+	ctx    context.Context
+}
+
+func NewKubeClientController(ctx context.Context,
+	useInClusterConfig bool, kubeconfig string) (*KubeClientController, error) {
```
Comment (rcanderson23) [inline internal/kubernetes/kubernetes.go:20]: Is this a linter making some of the funcs multiline? Don't often see that in Go.

**Review by rcanderson23 — COMMENTED** [review]
The `application` struct and related functions/structs feel like they should be in its own package which is called into with a `New` func and has something like `Serve` or `Run` to start it with the main func remaining in `cmd/teleport-challenge`.

File: internal/kubernetes/kubernetes.go (hunk 1 above)
Comment (jacobboykin) [reply internal/kubernetes/kubernetes.go:20]: I'm not sure if there's a linter out there that will do this. I did this manually to keep lines behind the vertical ruler guide in GoLand.

File: cmd/teleport-challenge/main.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	if err != nil {
+		return err
+	}
+
+	err = app.k8s.CheckServerConnection()
+	if err != nil {
+		return err
+	}
+
+	return nil
```
[the remark this answers is not in the document]
Comment (jacobboykin) [reply cmd/teleport-challenge/main.go]: Makes sense; fixed!

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+# =========================================================================
+# Config
+
+APP_NAME := teleport-challenge
+
+# =========================================================================
+# Run
+
+# Build and run the application via `go run` using the default flag values
+run-go:
```
[the remark this answers is not in the document]
Comment (jacobboykin) [reply Makefile:13]: Good catch; fixed!

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	go run ./cmd/${APP_NAME}
+
+# =========================================================================
+# Development
+
+# Install dependencies for Go development
+dev-install-dependencies:
```
[the remark this answers is not in the document]
Comment (jacobboykin) [reply Makefile]: I think homebrew works on linux. However, I haven't tested this, so I've updated the target name and description.

File: internal/kubernetes/kubernetes.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	}, nil
+}
+
+func (kube *KubeClientController) CheckServerConnection() error {
+	_, err := kube.client.ServerVersion()
+	if err != nil {
+		return err
+	}
+
+	return nil
```
[the remark this answers is not in the document]
Comment (jacobboykin) [reply internal/kubernetes/kubernetes.go:51]: I think this one should be left as is. If I changed the code to `return kube.client.ServerVersion()`, I'd have to change the function signature to `(kube *KubeClientController) CheckServerConnection() (*version.Info, error)` to deal with the multiple return values. At this time I don't think there's any need to return the K8s API server version info, so I'd rather drop that return value.

[the remark this answers is not in the document]
Comment (rcanderson23) [reply internal/kubernetes/kubernetes.go:51]: Ah, yeah. This one is just removing the `if err != nil` check though my wording made it confusing compared to the previous one. The function return an error so we don't need to check the error before returning.

[the remark this answers is not in the document]
Comment (jacobboykin) [reply internal/kubernetes/kubernetes.go]: Understood! Just fixed this

File: internal/json_helper/json_helper.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package json_helper
+
+import (
+	"encoding/json"
+	"net/http"
+)
+
+type Envelope map[string]interface{}
+
+func Write(w http.ResponseWriter, status int, data Envelope, headers http.Header) error {
```
Comment (kbence) [inline internal/json_helper/json_helper.go:10]: Nit: the name of the function is a bit confusing to me. It's doing a lot of things for a helper function named `Write`.

**Review by kbence — APPROVED** [review]
Looks good, I only had one small thing.

**Review by rcanderson23 — APPROVED** [review]

**Pull request merged**